src/framework/myCFG.ml

```diff
@@ -165,138 +165,214 @@ let createCFG (file: file) =
+          failwith "MyCFG.createCFG: unsupported stmt"
+    in
+    try
+      let initial_visited_sids = match parent with
+        | Some parent -> [parent.sid]
+        | None -> []
+      in
+      let r = find initial_visited_sids stmt in
+      if Messages.tracing then Messages.traceu "cfg" "-> %d\n" r.sid;
+      r
+    with Not_found ->
+      if Messages.tracing then Messages.traceu "cfg" "-> Not_found\n";
+      raise Not_found
   in
-  addCfg (Function dummy_func) (Ret (None, dummy_func), FunctionEntry dummy_func);
+  addEdge_fromLoc (FunctionEntry dummy_func) (Ret (None, dummy_func)) (Function dummy_func);
   (* We iterate over all globals looking for functions: *)
   iterGlobals file (fun glob ->
       match glob with
-      | GFun (fd,loc) ->
+      | GFun (fd, fd_loc) ->
         if Messages.tracing then Messages.trace "cfg" "Looking at the function %s.\n" fd.svar.vname;
+
+        if get_bool "dbg.cilcfgdot" then
+          Cfg.printCfgFilename ("cilcfg." ^ fd.svar.vname ^ ".dot") fd;
+
         (* Walk through the parameters and pre-process them a bit... *)
         do_the_params fd;
         (* Find the first statement in the function *)
-        let entrynode = realnode true (CF.getFirstStmt fd) in
+        let entrynode = find_real_stmt (CF.getFirstStmt fd) in
         (* Add the entry edge to that node *)
-        let _ = addCfg (Statement entrynode) ((Entry fd), (FunctionEntry fd)) in
+        addEdge (FunctionEntry fd) (fd_loc, Entry fd) (Statement entrynode);
         (* Return node to be used for infinite loop connection to end of function
          * lazy, so it's only added when actually needed *)
         let pseudo_return = lazy (
-          let newst = mkStmt (Return (None, loc)) in
+          let newst = mkStmt (Return (None, fd_loc)) in
           let start_id = 10_000_000_000 in (* TODO get max_sid? *)
-          let sid = Hashtbl.hash loc in (* Need pure sid instead of Cil.new_sid for incremental, similar to vid in Goblintutil.create_var. We only add one return stmt per loop, so the location hash should be unique. *)
+          let sid = Hashtbl.hash fd_loc in (* Need pure sid instead of Cil.new_sid for incremental, similar to vid in Goblintutil.create_var. We only add one return stmt per loop, so the location hash should be unique. *)
           newst.sid <- if sid < start_id then sid + start_id else sid;
           Hashtbl.add stmt_fundec_map newst.sid fd;
           let newst_node = Statement newst in
-          addCfg (Function fd) (Ret (None,fd), newst_node);
+          addEdge newst_node (fd_loc, Ret (None, fd)) (Function fd);
           newst_node
         )
         in
+        let loop_head_neg1 = NH.create 3 in
         (* So for each statement in the function body, we do the following: *)
         let handle stmt =
           (* Please ignore the next line. It creates an index of statements
            * so the Eclipse plug-in can know what function a given result
            * belongs to. *)
```

Comment:
We should update this comment, we no longer have an Eclipse plugin.
This is actually used by `analyses.ml` now (via `getFun` -> `get_containing_function`).

Reply:
Indeed, but I'll leave it for another PR because I want to consistently refactor all of these things like `MyCFG.getFun`, `Cilfacade.getFun`, `Cilfacade.getdec` and possibly something else I currently forgot. This shouldn't even be intertwined with CFG construction but be a separate pass that builds the helper mappings.

Reply:
I'll track this in #290.